src/processor.cpp

```diff
@@ -0,0 +1,394 @@
+    }
+    if ( !(fuel_intake >= 0) ) {
+        debugmsg("Invalid fuel intake \"%d\" in \"%s\".", fuel_intake, id.c_str());
+    }
+    if (!(duration > 0)) {
+        debugmsg("Invalid duration \"%d\" in \"%s\".", duration, id.c_str());
+    }
+}
+
+void processes::load(JsonObject &jo, const std::string &src)
+{
+    processes_data.load(jo, src);
+}
+
+void processes::check()
+{
+    processes_data.check();
+}
+
+void processes::reset()
+{
+    processes_data.reset();
+}
+
+// Processor namespace
+
+namespace
+{
+struct processor_data;
+using processor_id = string_id<processor_data>;
+
+struct processor_data {
+
+    processor_data() :
+        was_loaded( false ) {};
+
+    processor_id id; //furniture
+    furn_str_id next_type;
+    itype_id fuel_type;
+
+
+    std::vector<std::string> processes;
+
+    std::string full_message;
+    std::string empty_message;
+    std::string start_message;
+    std::string finish_message;
+    std::string processing_message;
+    std::string fail_message;
+
+    bool was_loaded;
+
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+
+};
+
+
+processor_id get_processor_id( const tripoint &pos )
+{
+    return processor_id( g->m.furn( pos ).id().str() );
+}
+
+generic_factory<processor_data> processors_data( "processor type", "id" );
+
+}
+
+void processor_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "next_type", next_type );
+    mandatory( jo, was_loaded, "fuel_type", fuel_type );
+    mandatory( jo, was_loaded, "processes", processes );
+
+    if( !was_loaded || jo.has_member( "messages" ) ) {
+        JsonObject messages_obj = jo.get_object( "messages" );
+
+        optional( messages_obj, was_loaded, "empty", empty_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "full", full_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "start", start_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "processing", processing_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "finish", finish_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "fail", fail_message, translated_string_reader );
+    }
+}
+
+void processor_data::check() const
+{
+    static const iexamine_function converter( iexamine_function_from_string( "converter" ) );
+    const furn_str_id processor_id( id.str() );
+
+    if( !processor_id.is_valid() ) {
+        debugmsg( "Furniture \"%s\" have no processor of the same name.", id.c_str() );
+    } else if( processor_id->examine != converter ) {
+        debugmsg( "Furniture \"%s\" can't act as a processor because the assigned iexamine function is not converter, but it has \"%s\" associated.",
+                  processor_id.c_str(), id.c_str() );
+    }
+    for( const std::string &elem : processes ) {
+        if( !get_process_id(elem).is_valid() ) {
```

Comment:
If you stored `elem` as an id, you could just use `elem.is_valid()` here, which would be cleaner.